src/download.rs

```diff
@@ -143,34 +154,54 @@ impl<'dfu, IO: DfuIo> ErasePage<'dfu, IO> {
         ),
         IO::Error,
     > {
-        let (page_size, rest_memory_layout) =
-            self.memory_layout.split_first().ok_or(Error::NoSpaceLeft)?;
+        let page_size = match self.dfu.protocol() {
+            DfuProtocol::Dfuse { memory_layout, .. } => {
+                let (page_size, rest_memory_layout) = memory_layout
+                    .as_ref()
+                    .split_first()
+                    .ok_or(Error::NoSpaceLeft)?;
+                log::trace!("Rest of memory layout: {:?}", rest_memory_layout);
+                *page_size
+            }
+            _ => return Err(Error::InCompatibleProtocol.into()),
+        };
         log::trace!("Page size: {}", page_size);
-        log::trace!("Rest of memory layout: {:?}", rest_memory_layout);
 
+        let (erased_pos, next_protocol) = match self.protocol {
+            ProtocolData::Dfuse {
+                address,
+                erased_pos,
+                address_set,
+            } => (
+                erased_pos,
+                ProtocolData::Dfuse {
+                    address,
+                    erased_pos: erased_pos
+                        .checked_add(page_size)
+                        .ok_or(Error::EraseLimitReached)?,
+                    address_set,
+                },
+            ),
+            _ => return Err(Error::InCompatibleProtocol.into()),
```

Comment:
I'm a bit in doubt about this. Should this error be a user error (an error type that the caller can handle) or a hard error (panic, unreachable).

It's okay to have a mix of panicking behavior and error handling. They both carry different meaning.

Another hint that gives me the impression that this should panic and not return an error is because the type Protocol is private and thus, the user cannot do anything at all to recover.

Reply:
I guess this could be an unreachable! instead indeed;

Reply:
managed to make this case impossible by refactoring things slightly so  neither returning an error or paniccing is needed anymore :)

Reply:
wow so cool

Reply:
was it due to the power of the state machine? xD I wonder (because I often feel using a state machine was a bit overengineered, it's so much simpler to write things just with simple code)

either way I'm happy with the change!

Reply:
It's thanks to making sure each step only has data they know it can actually handle; Which means there is no need to bail out if there is data they can't by definition.  That ability comes out from a state machine design indeed.

That said honestly i'm a bit on the fence about this particular state machine design and how e.g. it's consumed from the download function.. It does have an overenginered feel to it, but then on the flip side it makes everything quite explicit and leads to relatively simple code which is always great :)